src/Autofac.Pooling/RegistrationExtensions.cs

```diff
@@ -0,0 +1,369 @@
+        /// </para>
+        /// </summary>
+        /// <typeparam name="TLimit">Registration limit type.</typeparam>
+        /// <typeparam name="TActivatorData">Activator data type.</typeparam>
+        /// <typeparam name="TSingleRegistrationStyle">Registration style.</typeparam>
+        /// <param name="registration">The registration.</param>
+        /// <param name="poolPolicy">A custom policy for controlling pool behaviour.</param>
+        /// <param name="lifetimeScopeTags">Tags applied to matching lifetime scopes.</param>
+        /// <returns>The registration builder.</returns>
+        public static IRegistrationBuilder<TLimit, TActivatorData, TSingleRegistrationStyle>
+               PooledInstancePerMatchingLifetimeScope<TLimit, TActivatorData, TSingleRegistrationStyle>(
+                   this IRegistrationBuilder<TLimit, TActivatorData, TSingleRegistrationStyle> registration,
+                   IPooledRegistrationPolicy<TLimit> poolPolicy,
+                   params object[] lifetimeScopeTags)
+               where TSingleRegistrationStyle : SingleRegistrationStyle
+               where TActivatorData : IConcreteActivatorData
+               where TLimit : class
+        {
+            if (registration == null) throw new ArgumentNullException(nameof(registration));
+
+            RegisterPooled(registration, poolPolicy, lifetimeScopeTags);
+
+            return registration;
+        }
+
+        private static void RegisterPooled<TLimit, TActivatorData, TSingleRegistrationStyle>(
+            IRegistrationBuilder<TLimit, TActivatorData, TSingleRegistrationStyle> registration,
+            IPooledRegistrationPolicy<TLimit> registrationPolicy,
+            object[]? tags)
+            where TSingleRegistrationStyle : SingleRegistrationStyle
+            where TActivatorData : IConcreteActivatorData
+            where TLimit : class
+        {
+            if (registration == null) throw new ArgumentNullException(nameof(registration));
+
+            // Mark the lifetime appropriately.
+            var regData = registration.RegistrationData;
+
+            regData.Lifetime = new PooledLifetime();
+            regData.Sharing = InstanceSharing.None;
+
+            var callback = regData.DeferredCallback;
+
+            if (callback is null)
+            {
+                throw new NotSupportedException(RegistrationExtensionsResources.RequiresCallbackContainer);
+            }
+
+            if (registration.ActivatorData.Activator is ProvidedInstanceActivator)
+            {
+                // Can't use provided instance activators with pooling (because it would try to repeatedly activate).
+                throw new NotSupportedException(RegistrationExtensionsResources.CannotUseProvidedInstances);
+            }
+
+            var original = callback.Callback;
+
+            Action<IComponentRegistryBuilder> newCallback = registry =>
```

Comment:
It feels weird storing a reference to a local function outside of the declaring function though...I get that it's the same under the hood, but at least 'newCallback' is actually a variable I can pass around...you know?

Reply:
Reading up on it, it appears [there are some differences](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/local-functions) like the way exceptions are handled or the way parameters get passed. However, in this specific case, it may not matter. I'm fine leaving it (though I may later add a `[SuppressMessage]` attribute on it so it's not sitting in the "Problems" window in VS Code).